enable/disable profiles

bentiss opened the pull request:
enable and disable profiles in ratbagd, ratbagctl and ratbag-command.

The hidpp10 implementation was not done with that in mind, which explains the large commits for them.
The result is now less communications with the HID++ 1.0 devices as we are caching the data inside libratbag.

Hjdskes commented on ratbagd/ratbagd-profile.c:
```diff
@@ -325,6 +331,43 @@ static int ratbagd_profile_set_active(sd_bus_message *m,
+	(void) sd_bus_emit_signal(sd_bus_message_get_bus(m),
+				  RATBAGD_OBJ_ROOT,
+				  RATBAGD_NAME_ROOT ".Profile",
+				  "EnabledProfileChanged",
```
Just checking: is the intended usage here to check for `Enabled` in the callback of this signal? That'd be another round trip over DBus; perhaps we can emit a `ProfileEnabled`/`ProfileDisabled` signal instead.

bentiss replied:
Not a strong opinion. I can emit both if you want.

Hjdskes commented on tools/ratbagd.py:
```diff
@@ -374,6 +383,14 @@ def set_active(self):
+        return self._dbus_call("Enable", "")
+
+    def disable(self):
+        """Enable this profile."""
```
Nitpick: should be "Disable" here

bentiss replied:
good catch

Hjdskes commented on ratbagd/ratbagd-profile.c:
```diff
@@ -346,17 +389,35 @@ static int ratbagd_profile_get_resolution_by_index(sd_bus_message *m,
 	SD_BUS_PROPERTY("Leds", "ao", ratbagd_profile_get_leds, 0, 0),
 	SD_BUS_PROPERTY("ActiveResolution", "u", ratbagd_profile_get_active_resolution, 0, 0),
 	SD_BUS_PROPERTY("DefaultResolution", "u", ratbagd_profile_get_default_resolution, 0, 0),
+	SD_BUS_METHOD("Enable", "", "u", ratbagd_profile_enable, SD_BUS_VTABLE_UNPRIVILEGED),
```
Why not `SetEnabled(bool)` like we have for `SetActive`?

bentiss replied:
Because... lazy me?

bentiss commented:
> The changes made to ratbagd and the DBus interface look good to me, just have some questions regarding the intended use and some documentation fixes.

Just a quick FYI, @whot is currently looking at the dbus API and ideally we would suppress methods and only have properties. I did not think much about the API right now, I was just trying to have something that just works :)

I'll amend the changes, but beware that the dbus API will likely change in the (near) future.

Hjdskes commented:
I'm aware of this and am looking forward to it :)

whot commented:
that first commit (add missing Commit()) shows a slight problem we have with ratbagd. ratbagd returns the state of the device before the commit, so in the extreme case we can have a client change that state but never commits it to the device. The next client will then get a state that does not reflect the device. I wonder if we can do client tracking in ratbagd to reset the device on client disconnect.

whot commented on hwdb/70-libratbag-mouse.hwdb:
```diff
@@ -69,31 +69,39 @@ mouse:usb:v046dpc041:name:*:
+# http://support.logitech.com/en_us/product/g5-laser-mouse-product/specs
 mouse:usb:v046dpc049:name:*:
  RATBAG_DRIVER=hidpp10
  RATBAG_HIDPP10_DPI_LIST=400;800;1600;2000
```
is there something missing? If not, the commit msg should point out that those two are unknown, otherwise it looks like a copy/paste error.

bentiss replied:
IIRC, they are not unknown, they are just using the default value 1.
I guess we can still override this here.

whot commented on hwdb/70-libratbag-mouse.hwdb:
```diff
@@ -69,31 +69,39 @@ mouse:usb:v046dpc041:name:*:
 # G7
+# http://support.logitech.com/en_us/product/g7-laser-cordless-mouse/specs
 mouse:usb:v046dpc51a:name:*:
  RATBAG_DRIVER=hidpp10
```
as above

bentiss commented:
> that first commit (add missing Commit()) shows a slight problem we have with ratbagd. ratbagd returns the state of the device before the commit, so in the extreme case we can have a client change that state but never commits it to the device. The next client will then get a state that does not reflect the device. I wonder if we can do client tracking in ratbagd to reset the device on client disconnect.

Nah. I think we should consider the daemon state as being the correct one. If a client _forgets_ to call commit(), we should not have a different state for other clients. This will be race prone and we might end up having too much code for no benefit.
Anyway, if one client forgets to commit, the next one will more likely commit given that we expect clients to actually change the device.

> looks good, but i'd like this to be rebased on top of #242 please, sorry.

No worries. When I saw your local brnach, I already made up my mind that I would have to rewrite it.
I could also have dropped the ratbagd part, but that would just differ the commit in a separate PR, so no gain.

bentiss commented:
> looks good, but i'd like this to be rebased on top of #242 please, sorry.

Done. Squashed the 2 comments and also changed the `Enabled` property to be read-write.
